src/object.d

```diff
@@ -3241,6 +3241,40 @@ template RTInfo(T)
     enum RTInfo = null;
 }
 
+bool __equals(T1, T2)(T1[] lhs, T2[] rhs)
+{
+    if (lhs.length != rhs.length)
+        return 0;
+
+    import core.internal.traits : Unqual;
+    alias U1 = Unqual!T1;
+    alias U2 = Unqual!T2;
+    static assert(is(U1 == U2), "Internal error.");
+
+    static @trusted ref R at(R)(R[] r, size_t i) { return r.ptr[i]; }
+
+    //old path: Typeinfo_array => TypeInfo_struct
+    //static if (is(U1 == struct) || is(U1 == enum))
+    foreach (const u; 0 .. lhs.length)
+    {
+        auto s1 = at(lhs, u);
+        auto s2 = at(rhs, u);
+
+        static if (__traits(compiles, s1.opEquals(s2)))
+        {
+            auto result = (s1).opEquals(s2);
+            if (!result)
+                return result;
+        }
+        else
+        {
+            return s1.tupleof == s2.tupleof;
```

Comment:
Nice. I had a case where both ``` assert(a != b);``` and ```assert(a==b);``` passed with that code because I forgot to do this in the compiler:
```
                if (op == TOKnotequal)
                {
                    __equals = new NotExp(loc, __equals);
                }
```